---
title: index
weight: 10
---

```plain
$ lexicmap index -h
Generate an index from FASTA/Q sequences

Input:
 *1. Sequences of each reference genome should be saved in separate FASTA/Q files, with reference identifiers
     in the file names.
  2. Input plain or gzipped FASTA/Q files can be given via positional arguments or the flag -X/--infile-list
     with the list of input files.
  3. Input can also be a directory containing sequence files via the flag -I/--in-dir, with multiple-level
     sub-directories allowed. A regular expression for matching sequencing files is available via the flag
     -r/--file-regexp.
  4. Some none-isolate assemblies might have extremely large genomes (e.g., GCA_000765055.1, >150 mb).
     The flag -g/--max-genome is used to skip these input files, and the file list would be written to a file
     (-G/--big-genomes).

  Attention:
   *1) ► You can rename the sequence files for convenience because the genome identifiers in the index and
       search result would be: the basenames of files with common FASTA/Q file extensions removed, which
       are extracted via the flag -N/--ref-name-regexp.
       ► The extracted genome identifiers better be distinct/unique, which will be shown in search results
       and are used to extract subsequences in the command "lexicmap utils subseq".
    2) ► Unwanted sequences like plasmids can be filtered out by the name via regular expressions
       (-B/--seq-name-filter).

Important parameters:
  --- LexicHash mask generation ---
  0. -M/--mask-file,   ► File with custom masks, which could be generated by "lexicmap utils gen-masks".
                       This flag oversides -k/--kmer, -m/--masks, -s/--rand-seed, -n/--top-n, and -P/--prefix-ext.
  1. -k/--kmer,        ► K-mer size (maximum: 32, default: 31).
                       ► Bigger values improve the search specificity and do not increase the index size.
  2. -m/--masks,       ► Number of masks (default: 40000).
                       ► Bigger values improve the search sensitivity and increase the index size.
  3. -n/--top-n,       ► The top N largest genomes for generating mask (default 20).
                       ► Bigger values increase the indexing memory occupation (1.0~1.5 GB per genome for 20k masks).
  4. -P/--prefix-ext,  ► Extension length of prefixes (default 8).
                       ► Bigger values improve the search sensitivity by decreasing the maximum seed distances.

  --- seeds (k-mer-value data) ---
  1. -c/--chunks,      ► Number of seed file chunks (maximum: 128, default: #CPUs).
                       ► Bigger values accelerate the search speed at the cost of a high disk reading load.
                       The maximum number should not exceed the maximum number of open files set by the
                       operating systems.
  2. -p/--partitions,  ► Number of partitions for indexing each seed file (default: 512).
                       ► Bigger values bring higher memory occupation. 512 is a good value with high searching speed,
                       Larger or smaller values would decrease the speed in "lexicmap search".
                       ► After indexing, "lexicmap utils reindex-seeds" can be used to reindex the seeds data with
                       another value of this flag.
  3. --max-open-files, ► Maximum number of open files (default: 512).
                       ► It's only used in merging indexes of multiple genome batches.

  --- genome data ---
  1. -b/--batch-size,  ► Maximum number of genomes in each batch (maximum: 131072, default: 10000).
                       ► If the number of input files exceeds this number, input files are into multiple
                       batches and indexes are built for all batches. Next, seed files are merged into a
                       big one, while genome data files are kept unchanged and collected.
                       ► Bigger values increase indexing memory occupation, while improving the search speed.

Usage:
  lexicmap index [flags] [-k <k>] [-m <masks>] { -I <seqs dir> | -X <file list>} -O <out dir>

Flags:
  -b, --batch-size int            ► Maximum number of genomes in each batch (maximum value: 131072)
                                  (default 10000)
  -G, --big-genomes string        ► Out file of skipped files with genomes >= -G/--max-genome
  -c, --chunks int                ► Number of chunks for storing seeds (k-mer-value data) files.
                                  (default 16)
  -r, --file-regexp string        ► Regular expression for matching sequence files in -I/--in-dir,
                                  case ignored. (default "\\.(f[aq](st[aq])?|fna)(.gz)?$")
      --force                     ► Overwrite existing output directory.
  -h, --help                      help for index
  -I, --in-dir string             ► Directory containing FASTA/Q files. Directory symlinks are followed.
  -k, --kmer int                  ► Maximum k-mer size. K needs to be <= 32. (default 31)
  -M, --mask-file string          ► File of custom masks. This flag oversides -k/--kmer, -m/--masks,
                                  -s/--rand-seed, --prefix, -n/--top-n, and -P/--prefix-ext.
  -m, --masks int                 ► Number of masks. (default 40000)
  -g, --max-genome int            ► Maximum genome size. Extremely large genomes (non-isolate
                                  assemblies) will be skipped. (default 15000000)
      --max-open-files int        ► Maximum opened files, used in merging indexes. (default 512)
  -O, --out-dir string            ► Output directory.
  -p, --partitions int            ► Number of partitions for indexing seeds (k-mer-value data) files.
                                  (default 512)
  -P, --prefix-ext int            ► Extension length of prefixes, higher values -> smaller maximum
                                  seed distances. (default 8)
  -s, --rand-seed int             ► Rand seed for generating random masks. (default 1)
  -N, --ref-name-regexp string    ► Regular expression (must contains "(" and ")") for extracting the
                                  reference name from the filename. (default
                                  "(?i)(.+)\\.(f[aq](st[aq])?|fna)(.gz)?$")
      --save-seed-pos             ► Save seed positions, which can be inspect with "lexicmap utils
                                  seed-positions".
  -B, --seq-name-filter strings   ► List of regular expressions for filtering out sequences by
                                  header/name, case ignored.
  -n, --top-n int                 ► Select the top N largest genomes for generating masks. (default 20)
```
